cloudinit/distros/__init__.py

```diff
@@ -818,20 +823,108 @@ def create_user(self, name, **kwargs):
-        # Set password if plain-text password provided and non-empty
-        if "plain_text_passwd" in kwargs and kwargs["plain_text_passwd"]:
-            self.set_passwd(name, kwargs["plain_text_passwd"])
-
-        # Set password if hashed password is provided and non-empty
-        if "hashed_passwd" in kwargs and kwargs["hashed_passwd"]:
-            self.set_passwd(name, kwargs["hashed_passwd"], hashed=True)
+        pre_existing_user = not self.add_user(name, **kwargs)
+
+        blank_password = False
+        has_password = False
+        password_key = None
+
+        if "plain_text_passwd" in kwargs:
+            has_password = True
+            password_key = "plain_text_passwd"
+            if kwargs["plain_text_passwd"]:
+                # Set password if plain-text password provided and non-empty
+                self.set_passwd(name, kwargs["plain_text_passwd"])
+            else:
+                blank_password = True
+
+        if "hashed_passwd" in kwargs:
+            has_password = True
+            password_key = "hashed_passwd"
+            if kwargs["hashed_passwd"]:
+                # Set password if hashed password is provided and non-empty
+                self.set_passwd(name, kwargs["hashed_passwd"], hashed=True)
+            else:
+                blank_password = True
+
+        if pre_existing_user:
+            if not has_password:
+                if "passwd" in kwargs:
+                    password_key = "passwd"
+                    # Only "plain_text_passwd" and "hashed_passwd"
+                    # are valid for an existing user.
+                    LOG.warning(
+                        "'passwd' in user-data is ignored for existing "
+                        "user %s",
+                        name,
+                    )
 
-        # Default locking down the account.  'lock_passwd' defaults to True.
-        # lock account unless lock_password is False.
+                # As no password set for the existing user in user-data then
+                # check if the existing user's hashed password value in
+                # /etc/shadow is blank (whether locked or not).
+                cmd = [
+                    "grep",
+                    "-q",
+                    "-e",
+                    "^%s::" % name,
+                    "-e",
+                    "^%s:!:" % name,
+                    "/etc/shadow",
+                ]
+                try:
+                    subp.subp(cmd)
+                except subp.ProcessExecutionError as e:
+                    if e.exit_code == 1:
+                        # Exit code 1 means 'grep' didn't find empty password
+                        has_password = True
+                    else:
+                        util.logexc(
+                            LOG,
+                            "Failed to Check if user %s already has "
+                            "a password",
+                            name,
+                        )
+                        raise e
```

Comment:
Let's make a encapsulate this logic in an instance method `is_user_password_empty_or_locked()` to make it easier to digest. Then that method can grep against `shadow_files = [f for f in [self.shadow_fn, self.extrausers_shadow_fn] if os.path.exists(f)]`

Reply:
@blackboxsw

Yeah I'm going to move this logic into a function, especially as I just realised that it doesn't cater for the BSDs - FreeBSD uses /etc/master.passwd instead of /etc/shadow, I need to work out what the other BSDs use.

Reply:
+1

Reply:
@blackboxsw

logic moved into function as discussed. Also now caters for NetBSD/FreeBSD/OpenBSD behaviour correctly (I did some, non cloud-init, testing with VMs of those).